src/_nebari/upgrade.py

```diff
@@ -1297,11 +1298,34 @@ def _version_specific_upgrade(
 
             urllib3.disable_warnings()
 
+            keycloak_username = os.environ.get("KEYCLOAK_ADMIN_USERNAME", "root")
+            keycloak_password = os.environ.get(
+                "KEYCLOAK_ADMIN_PASSWORD",
+                config["security"]["keycloak"]["initial_root_password"],
+            )
+
             keycloak_admin = get_keycloak_admin(
                 server_url=f"https://{config['domain']}/auth/",
-                username="root",
-                password=config["security"]["keycloak"]["initial_root_password"],
+                username=keycloak_username,
+                password=keycloak_password,
             )
+
+            try:
+                # Quick test to connect to Keycloak
+                keycloak_admin.get_server_info()
+            except ValueError as e:
+                if "invalid_grant" in str(e):
+                    print(
+                        "[red bold]Failed to connect to the Keycloak server.[/red bold]\n"
+                        "[yellow]This may occur if the default admin credentials have been changed for security reasons.[/yellow]\n"
+                        "Please update the [bold]KEYCLOAK_ADMIN_USERNAME[/bold] and [bold]KEYCLOAK_ADMIN_PASSWORD[/bold] environment variables with the new credentials and try again."
+                    )
+                    exit()
+                else:
+                    # Handle other exceptions
+                    print(f"[red bold]An unexpected error occurred: {e}[/red bold]")
```

Comment:
Will the whole stack trace be in red bold?

Reply:
I see it doesn't print out the stack trace.  I agree with the comments on https://stackoverflow.com/a/1483488/9848141 that repr would be better than str(e) to give more info about what is happening.